modules/post/multi/gather/enum_software_versions.rb

```diff
@@ -0,0 +1,139 @@
+      end
+      file = store_loot('host.windows.software.versions', 'text/plain', session, listing, 'installed_software.txt', 'Installed Software and Versions')
+      print_good("Stored information about the installed products to the loot file at #{file}")
+    when 'linux'
+      # All of the following options were taken from https://distrowatch.com/dwres.php?resource=package-management
+      operating_system = cmd_exec('hostnamectl')
+      if operating_system =~ /not found/
+        print_error("Can't determine the host's OS, as the 'hostnamectl' command was not found")
+        return nil
+      elsif operating_system =~ /ubuntu/i || operating_system =~ /debian/i || operating_system =~ /elementary/i || operating_system =~ /mint/i || operating_system =~ /MX/ || operating_system =~ /zorin/i || operating_system =~ /kali/i
+        listing = cmd_exec('apt list --installed')
+        if listing =~ /not found/
+          print_error("The command 'apt' does not exist on the host")
+          return nil
+        end
+        unless listing =~ /listing\.\./i
+          print_error('Was unable to dump the versions of software installed!')
+          return nil
+        end
+        store_linux_loot(listing)
+      elsif operating_system =~ / [aA]rch / || operating_system =~ /manjaro/i
+        listing = cmd_exec('pacman -Q')
+        if listing =~ /not found/
+          print_error("The command 'pacman' does not exist on the host")
+          return nil
+        end
+        store_linux_loot(listing)
+      elsif operating_system =~ /opensuse/i
+        listing = cmd_exec('zypper search -is')
+        if listing =~ /not found/
+          print_error("The command 'zypper' does not exist on the host")
+          return nil
+        end
+        store_linux_loot(listing)
+      elsif operating_system =~ /fedora/i || operating_system =~ /centos/i
+        listing = cmd_exec('rpm -qa')
+        if listing =~ /not found/
+          print_error("The command 'rpm' does not exist on the host")
+          return nil
+        end
+        store_linux_loot(listing)
+      elsif operating_system =~ /alpine/i
+        listing = cmd_exec('apk info')
+        if listing =~ /not found/
+          print_error("The command 'apk' does not exist on the host")
+          return nil
+        end
+        store_linux_loot(listing)
+      elsif operating_system =~ /gentoo/i
+        listing = cmd_exec('qlist -i')
+        if listing =~ /not found/
+          print_error("The command 'qlist' does not exist on the host")
+          return nil
+        end
+        store_linux_loot(listing)
+      elsif operating_system =~ /freebsd/i
+        listing = cmd_exec('pkg info')
+        if listing =~ /not found/
+          print_error("The command 'pkg' does not exist on the host")
+          return nil
+        end
+        store_linux_loot(listing)
+      else
+        print_error("The target's operating system is not supported at this time.")
+        return nil
+      end
```

Comment:
For all of these links ones, how about reducing the repetition by using a pattern like:
```ruby
if operating_system =~ /ubuntu/i
  cmd = %w{ apt list --installed }
elsif operating_system =~ /freebsd/i
  cmd = %w{ pkg inf }
elsif ...
  # ...
else
  print_error("The target's operating system is not supported at this time.")
  return nil
end

if (listing = cmd_exec(cmd.join(' '))) =~ /not found/
  print_error("The command '#{cmd[0]}' does not exist on the host")
else
  store_linux_loot(listing)
end```

That would go a long way in reducing the repition.

Reply:
Thanks @smcintyre-r7 I was actually thinking of trying to reduce this code myself but wasn't sure how to best approach it; your way makes perfect sense. Let me go ahead and implement that so long; appreciate you pointing this out 👍

Reply:
Next commit should incorporate the changes recommended here.

Reply:
I suspect matching `not found` may end poorly depending on OS locale.

The `Post::Common` mixin supports a `command_exists?` function to determine whether a command exists (in `$PATH`). This is preferred. 

```
# grep -rn "def command_exists" lib/ -A 10
lib/msf/core/post/common.rb:242:  def command_exists?(cmd)
lib/msf/core/post/common.rb-243-    if session.platform == 'windows'
lib/msf/core/post/common.rb-244-      # https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/where_1
lib/msf/core/post/common.rb-245-      # https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/if
lib/msf/core/post/common.rb-246-      cmd_exec("cmd /c where /q #{cmd} & if not errorlevel 1 echo true").to_s.include? 'true'
lib/msf/core/post/common.rb-247-    else
lib/msf/core/post/common.rb-248-      cmd_exec("command -v #{cmd} && echo true").to_s.include? 'true'
lib/msf/core/post/common.rb-249-    end
lib/msf/core/post/common.rb-250-  rescue
lib/msf/core/post/common.rb-251-    raise "Unable to check if command `#{cmd}' exists"
lib/msf/core/post/common.rb-252-  end
```

Reply:
Thanks @bcoles, thats a fair point, let me see if I can't use `command_exists?` instead, my way was a quick hack I put together and I agree its not always guaranteed to return `not found` as a string.

Reply:
Ok implemented this throughout the code, and this should now be resolved in the next commit.